Index: src/tools/save_bag_data.cpp
===================================================================
--- src/tools/save_bag_data.cpp	(revision 46820)
+++ src/tools/save_bag_data.cpp	(working copy)
@@ -126,7 +126,7 @@
     	printf("Saving data\n");
 
     	sprintf(name, "%.4d_cloud.pcd", index);
-    	pcl::io::savePCDFile(name, *point_cloud, true);
+    	pcl::io::savePCDFile(name, *point_cloud, Eigen3::Vector4f::Zero (), Eigen3::Quaternionf::Identity (), true);
     	sprintf(name, "%.4d_cloud.msg", index);
     	serialize(name, *point_cloud);
 
